ASoC: SOF: ipc4: Add topology fallback support for CPC information for modules

ujfalusi opened the pull request:
Hi,

replaces #4262 as for some reason that PR got stuck in error.

the MOD_INIT_INSTANCE message always contains a base_cfg struct and in there the CPC value must be set to non 0 to indicate the processing need of the instantiated module with ti's configuration.
Currently we have one module specific token to pass a single, static CPC value in topology. Some module has this set, some does not, but in any case it is not reflecting the real processing need based on the configuration, parameters.

The firmware's manifest however contains mod_cfg array which contains CPC value for the module's different configuration, the CPC number should be looked up by matching the obs/ibs in manifest mod_cfg and in the runtime base_cfg obs/ibs and pick the exact matched CPC or maximum CPC.

Since the manifest is part of the firmware binary, it is hard to modify and might contain out of date or non correct value, hence we should have a way to provide similar mod_cfg type of info via tplg and use that if it is present.

This series implements this CPC lookup and configuration for the modules with the following rule:

The CPC value will be set on the first matching case of:
CPC is set in topology for the module as a single override value
CPC can be found in tplg mod_cfg with exact match based on obs/ibs
CPC can be found in manifest mod_cfg with exact match based on obs/ibs
Maximum CPC value for the module from topology mod_cfg
Maximum CPC value for the module from manifest mod_cfg
CPS value of the platform (iow, the frequency)
If we still have 0 as CPC, assume a core frequency of 400MHz and use that as CPC.

A custom 'DSP frequency boost' kcontrol is provided to force the DSP to run at maximum frequency in user space request.

ujfalusi committed:
ASoC: SOF: ipc4-topology: Rename available_audio_format to be more generic

Currently we are using the term available_audio_format for structures which
only contains the module's available input and output formats.
This can be seen as a common (as it is common among  all module types)
module parameters storage.

In order to be able to extend the generic/common module information, rename
the sof_ipc4_available_audio_format to sof_ipc4_module_params.

Rename the (linux only) token identifier from SOF_AUDIO_FMT_NUM_TOKENS to
SOF_MODULE_PARAM_NUM_TOKENS.

No change in functionality apart from the renaming.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Add tokens for IPC4 module configuration

Add token information for OBS, IBS, CPC as module configuration parameters
and a token for the number of configuration entries for the module.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Use tplg module configuration for CPC setup

The topology file can contain a mod_cfg array with obs/ibs/cpc numbers to
override the information present in the manifest.

This patch adds the code to query the mod_cfg from topology and use it if
needed.

The CPC value will be set on the first matching case of:
CPC is set in topology for the module as a single override value
CPC can be found in tplg mod_cfg with exact match based on obs/ibs
CPC can be found in manifest mod_cfg with exact match based on obs/ibs
Maximum CPC value for the module from topology mod_cfg
Maximum CPC value for the module from manifest mod_cfg

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-loader: Save the maximum core frequency value

The core frequency can be used as a fallback in case a module does not
have valid CPC number assigned.
In this case the maximum core frequency will force the DSP to it's maximum
speed to avoid CPU starvation.

It can be also used as a 'boost' value to force the DSP to
run at maximum speed with REGISTER_KCPS large config message.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4-topology: Use max_core_frequency as fallback for CPC

If we cannot retrieve non 0 value for the module's CPC, use the
platforms's maximum frequency number to force the DSP to full speed
to avoid underclocking it.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
@ranj063 , I have addressed your comments with this PR.

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,53 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+/**
```
@ujfalusi the IBS*obs/1000 comment is still there in the commit message?

ujfalusi replied:
Oh, I missed this, only reworded the other patch.

ujfalusi commented:
Changes since v1:
- Removed the remaining "ibs*obs/1000" mention from commit message

keqiaozhang commented:
SOFCI TEST

RanderWang commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1231,6 +1231,23 @@ sof_ipc4_init_base_module_cfg(struct snd_sof_dev *sdev,
+		if (ipc4_data->max_kcps)
+			base_config->cpc = ipc4_data->max_kcps * 1000;
+		else
+			base_config->cpc = 400000000;
```
We can query HW clk by get_large_config IPC4_FAST_CLOCK_FREQ_HZ_FW_CFG

ujfalusi replied:
Oh, I did not knew that there is such a thing, then I can drop all the max_kcps addition also and only query it from the fw.

RanderWang reviewed (commented):
Can we use hw clk queried from FW, Not 400 in kernel ?

ujfalusi commented:
Changes since v2:
- Use the max core frequency from the fw_params reply (suggested by @RanderWang)
- Drop the per platform setting of max_kcps.

kv2019i commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -81,6 +82,7 @@ struct sof_ipc4_fw_data {
 	u32 mtrace_log_bytes;
 	int max_num_pipelines;
 	u32 max_libs_count;
+	u32 max_core_frequency;
```
@ujfalusi No 4.3+Ghz DSP support? :) Oh well, the IPC4 interface is limited to u32 already, so probably safe to go with this (famous last words)

paulstelian97 replied:
Imagine this trips when we run SOF with Zephyr on a CPU core (such as on i.MX93) and the core somehow shoots up above that (it doesn't for this specific case). Future ARM based boards which do this separate-core thingy where this can be an issue?

CC: @LaurentiuM1234

ujfalusi replied:
It is a u32 in the received message, I think a new tuple could be added for representing it in MHz, KHz?

kv2019i commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1231,6 +1231,16 @@ sof_ipc4_init_base_module_cfg(struct snd_sof_dev *sdev,
 	/* Update the base_config->cpc from the module manifest - max CPC value */
```
typo in commit: "use the paltform's" -> "use the platform's"

ujfalusi replied:
Oh, thanks, I will update.

kv2019i reviewed (dismissed):
Good to go!

ujfalusi commented:
Change since v3:
- fix typo in commit message of the last patch

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -28,11 +28,13 @@ enum sof_ipc4_mtrace_type {
  */
 struct sof_ipc4_fw_module {
 	struct sof_man4_module man4_module_entry;
+	const struct sof_man4_module_config *fw_mod_cfg;
```
commit message: I couldn't follow the wording in quotes:


Save a pointer to the firmware module configuration area in the manifest
"for each module which has at least one for later use"

at least one what?

ujfalusi replied:
Sure, I reword it: If the module have mod_cfg we save it for later use. The has or not has does not give any added information, we save the mod_cfg of the module, if it has, we have a pointer if it does not have then we have a NULL.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,53 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+		if (basecfg->obs == fw_mod_cfg[i].obs &&
+		    basecfg->ibs == fw_mod_cfg[i].ibs &&
+		    cpc_pick < fw_mod_cfg[i].cpc)
+			cpc_pick = fw_mod_cfg[i].cpc;
```
cpc_pick is initialized to zero and is an u32, it will by construction always be smaller than any other value.

This initialization will never be performed.

ujfalusi replied:
I'm not following you.
fw_mod_cfg[i].cpc is also u32

cpc_pick = 0;
fw_mod_cfg[0].cpc = 0;
fw_mod_cfg[1].cpc = 10;
fw_mod_cfg[2].cpc = 5;
fw_mod_cfg[3].cpc = 50;

cpc_pick will be 50 after the loop:
i = 0;
  cpc_pick == fw_mod_cfg[i].cpc -> no change in cpc_pick
i = 1;
  cpc_pick < fw_mod_cfg[i].cpc -> cpc_pick = 10;
i = 2;
  cpc_pick > fw_mod_cfg[i].cpc -> no change in cpc_pick
i = 3;
  cpc_pick < fw_mod_cfg[i].cpc ->cpc_pick = 50;

lyakh replied:
agree with @ujfalusi , @plbossart could you double-check?

plbossart replied:
never mind, I read this sideways. stupid me.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,53 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+			max_cpc = fw_mod_cfg[i].cpc;
+	}
+
+	if (!cpc_pick) {
```
always true test.

ujfalusi replied:
No, it is not, if we have a matching ibs/obs with non 0 cpc, the cpc_pick is not going to be 0

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -423,6 +423,53 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
+	for (i = 0; i < fw_module->man4_module_entry.cfg_count; i++) {
+		if (basecfg->obs == fw_mod_cfg[i].obs &&
+		    basecfg->ibs == fw_mod_cfg[i].ibs &&
+		    cpc_pick < fw_mod_cfg[i].cpc)
```
you really need a comment to explain why we need to look at obs/ibs while selecting a cpc value. It's far from obvious to me.

ujfalusi replied:
It is a best effort pick based on the module's external configuration combination as the numbers are recorded in the mod_cfg in relation to these.
I'll add a comment.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1032,6 +1077,52 @@ static int sof_ipc4_update_hw_params(struct snd_sof_dev *sdev, struct snd_pcm_hw
+		if (basecfg->obs == mod_cfgs[i].obs &&
+		    basecfg->ibs == mod_cfgs[i].ibs &&
+		    cpc_pick < mod_cfgs[i].cpc)
+			cpc_pick = mod_cfgs[i].cpc;
```
I've seen that code before and it's still wrong - and it's not clear why we copy/pasted. can a helper be used?

ujfalusi replied:
We have different struct for manifest_mod_cfg and for the tplg_mod_cfg.
The tplg version only contains ibs/obs/cpc, while the manifest have more fields (is_bytes/cps/ibs/obs/module_flags/cpc/obls)

ujfalusi replied:
I still thin the code is OK.

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -391,6 +391,13 @@ int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
+			ipc4_data->max_core_frequency = *tuple->value;
+			/* Assume 400MHz if the received value is 0 */
+			if (!ipc4_data->max_core_frequency)
+				ipc4_data->max_core_frequency = 400000000;
```
we shouldn't do this. This is generic code, the max_core_frequency should be in the chip descriptors or something private to each vendor/SOC.

ujfalusi replied:
So, update the sof_dev_desc to contain max_core_frequency?
or rather add a new optional callback to snd_sof_dsp_ops (`get_dsp_max_frequency()`) and use that?

plbossart replied:
Yes, something coming from the platform.

I am however confused by all this, we do have cases for Intel where we lower the clock to use e.g. 120 MHz. In that case, what is the relevance of the max_core_frequency? We should not assume that the cores always run flat out at the max frequency.

Not that I think of if, why do we even need a concept of max frequency?

kv2019i replied:
@plbossart This is just used as a fallback in case no guidance is given at what frequency to run (either from fw manifest or from topology). This fallback is same as the behaviour has been in all past SOF releases, e.g. run at max speed of DSP and clock-gate. In IPC4, driver HAS to provide this information, otherwise DSP will run at the lowest possible clock and that typically will not be enough and you will get errors. In IPC3, driver did not have to provide this information, so we didn't have the problem. So in summary, if you care about the DSP speed, you MUST provide correct data via manifest or topology. Otherwise DSP will run at the fast clock always.

plbossart replied:
@kv2019i I don't fully understand why the manifest or topology would tell us as which frequency to run. That'd be either an optimistic or worst-case value, in real life one would need to track usages and set the frequency as needed with some sort of governor.

It seems like we are in the Neanderthal age of frequency scaling...Dare I say Jurassic age?

kv2019i replied:
@plbossart Don't shoot the messenger :) That is how the IPC4 FW operates and if we don't provide values, IPC4 firmwares will run at 38.4Mhz in all scenarios.

plbossart replied:
@kv2019i I have lost track of how IPC4 is supposed to work. In the early days the firmware was supposed to be minimal and stupid, and only apply the decisions made by the mighty host driver. Now it seems the firmware can decide at which frequency to run based on module configurations? 

I am not getting the full picture on who does what....

kv2019i replied:
@plbossart I'll loop in @abonislawski who is implementing this on FW side. 
I think that "early days" philosophy still stands. FW doesn't have any policy, but just acts on the data the mighty host driver sends. Host tells you need X MCPS to run pipeline A (when creating pipeline A), and FW will bump freq to meet that. If host sets up two instances of pipeline A, FW needs to bump to 2X and so forth.

plbossart replied:
@kv2019i can you remind me how the host keeps track of the total X MCPS needed to run pipeline A? I've lost the plot, and I still don't get what the concept of pipeline plays here given that modules are be assigned to different cores. The pipeline was always notional for IPC4, modules are what matters.

ujfalusi replied:
@plbossart, we have the CPC value per modules, when a module is initialized we send this CPC, the firmware  adds these values together and that gives the whole CPC need for the initialized modules.
When the module is removed, it's CPC is also removed from the sum.

With the CPC the kernel is not directly telling the firmware to pick a certain freq, it just gives the firmware the info that a certain component needs X cycles to run, sum up all the X cycles and you get the frequency need.

The fallback to platform max is really  a fallback to ensure that we will not underclock the DSP, it is better to run it full speed than have audio artifacts and errors due to starvation.

ujfalusi committed:
ASoC: SOF: Remove unused set_clk / snd_sof_dsp_set_clk ops

The snd_sof_dsp_set_clk() is not used and the callback is not used by any
platform.

Remove them.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Add new ops to query the DSP max frequency to snd_sof_dsp_ops

The new, optional callback can be used by the core to query the maximum
core frequency of the DSP.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: hda: Add callback to retrieve max core frequency

The DSP in all HDA platforms are running at 400 MHz, add a generic callback
to return this value unconditionally.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v4:
- Commit message for patch 2 updated
- Comments added to explain the CPC lookup decision
- New ops added to retrieve the maximum DSP frequency and implemented for Intel HDA platforms (returning 400MHz)

plbossart commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -204,6 +205,11 @@
+#define SOF_TKN_MOD_CFG_IBS			1990
+#define SOF_TKN_MOD_CFG_OBS			1991
+#define SOF_TKN_MOD_CFG_CPC			1992
+
```
follow-up patch: maybe we should move the AMD stuff below to a different location to keep tokens in order.

plbossart replied:
still open

ujfalusi replied:
OK

plbossart commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -204,6 +205,11 @@
 
+/* module configurations */
+#define SOF_TKN_MOD_CFG_IBS			1990
+#define SOF_TKN_MOD_CFG_OBS			1991
```
we seem to always assume a module has both an input and output buffer.

What about e.g. a tone generator, it wouldn't have an input buffer?

plbossart replied:
still open

ujfalusi replied:
Right, that is a topology limitation, in a way.
If the module is output only then it's mod_cfg is:
```
Object.Base.mod_cfg.1 {
	obs 256
	cpc 1456
}

Object.Base.mod_cfg.2 {
	obs 384
	cpc 4242
}

Object.Base.mod_cfg.3 {
	obs 256
	cpc 2566
}
```
The `ibs` of the mod_cfg will be 0.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1032,6 +1077,64 @@ static int sof_ipc4_update_hw_params(struct snd_sof_dev *sdev, struct snd_pcm_hw
+	 * Find the best matching (highest) CPC value based on the module's
+	 * external configuration (IBS/OBS).
+	 * The CPC value in each module config entry has been measured and
+	 * recorded as a IBS/OBS/CPC triplet.
```
what if there is no IBS or no OBS? Why would we assume that this is always a triplet?

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -287,6 +287,9 @@ struct snd_sof_dsp_ops {
 			       const struct sof_dsp_power_state *target_state); /* optional */
 
+	/* DSP clocking */
+	unsigned long (*get_max_frequency)(struct snd_sof_dev *sof_dev); /* optional */
```
what is the return value? Hz? kHz? MHz?

plbossart replied:
still open

ujfalusi replied:
In Hz, I'll update the comment.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1776,6 +1776,12 @@ void hda_unregister_clients(struct snd_sof_dev *sdev)
+unsigned long hda_get_max_frequency(struct snd_sof_dev *sdev)
+{
+	/* The DSP is running at 400 MHz in all HDA platforms */
+	return 400000000;
```
not exactly 400 MHz, that's not completely right

plbossart replied:
still open.

ujfalusi replied:
The firmware also said that it is 400MHz
```
# git grep CLK_MAX_CPU_HZ

src/platform/meteorlake/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	400000000
src/platform/tigerlake/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	400000000

# git grep CLK_MAX_CPU_HZ v2.2

v2.2:src/platform/apollolake/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	400000000
v2.2:src/platform/baytrail/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	343000000
v2.2:src/platform/cannonlake/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	400000000
v2.2:src/platform/haswell/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	320000000
v2.2:src/platform/icelake/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	400000000
v2.2:src/platform/suecreek/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	400000000
v2.2:src/platform/tigerlake/include/platform/lib/clk.h:#define CLK_MAX_CPU_HZ	400000000
```

plbossart commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -391,6 +391,27 @@ int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
+					dev_warn(sdev->dev,
+						"out of range frequency: %lu\n",
+						max_freq);
+					ipc4_data->max_core_frequency = U32_MAX;
```
that sounds extreme!

ujfalusi replied:
As NXP commented, they might have 4294967295Hz+ when they run SOF on an ARM core.
In any case, we are saving an `unsigned long` as `u32`, it is better to be safe than sorry later.

See: https://github.com/thesofproject/linux/pull/4315/commits/e005ce8e4fbfd3064163dd81714db9142a378c35#r1175180853

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1243,6 +1243,16 @@ sof_ipc4_init_base_module_cfg(struct snd_sof_dev *sdev,
+	if (!base_config->cpc) {
+		struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+
+		base_config->cpc = ipc4_data->max_core_frequency;
```
What is the definition of CPC? Cycles per chunk?

This looks like CPS (cycles per second).

Different but related values.

plbossart reviewed (changes requested):
I am having a bit of allergies to the concept that all modules would have ibs/obs - that's not true in the general case. Also the units are not clear for frequencies and cpc/cps.

ujfalusi commented:
Changes since v5:
- Rebased on topic/sof-dev due to complicated conflict brought in by #4312
- The CPC configuration is moved to `sof_ipc4_update_resource_usage()` which is the renamed `sof_ipc4_update_pipeline_mem_usage()`

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -968,6 +968,14 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+	sof_ipc4_update_cpc_from_manifest(sdev, swidget->module_info, base_config);
+
+	dev_dbg(sdev->dev, "ibs / obs / cpc for %s: %u / %u / %u\n",
+		swidget->widget->name, base_config->ibs, base_config->obs,
```
I still don't get how we deal with a module that does NOT have an input buffer or pin.

ujfalusi replied:
Then the `ibs` is 0, right?

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -939,8 +984,67 @@ static void sof_ipc4_widget_free_comp_process(struct snd_sof_widget *swidget)
+	 * Find the best matching (highest) CPC value based on the module's
+	 * external configuration (IBS/OBS).
+	 * The CPC value in each module config entry has been measured and
+	 * recorded as a IBS/OBS/CPC triplet.
```
what if there is not ibs?

or what if there is no obs?

The assumption that all modules have input and output buffers does not make any sense to me.

ujfalusi replied:
If no ibs, then ibs is 0, right?
I don't see any problem with that.
The array in manifest is created based on the measured configuration, so ibs=0, the module have ibs=0, so that part is constant.

plbossart commented on sound/soc/sof/ipc4-topology.h:
```diff
@@ -166,18 +166,39 @@ struct sof_ipc4_pin_format {
+ * struct sof_ipc4_tplg_mod_cfg - module configuration entry from topology
+ * @ibs: Input Buffer Size
+ * @obs: Output Buffer Size
+ * @cpc: Cycles Per Chunk
```
what is a chunk

ujfalusi replied:
a chunk that is processed at once, typically 1ms.

But we don't really care in kernel, we read the number from manifest/topology and pass it to the firmware and we trust that they are added correctly.
We don't verify the ibs/obs numbers either for that matter.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1776,6 +1776,12 @@ void hda_unregister_clients(struct snd_sof_dev *sdev)
 	hda_probes_unregister(sdev);
 }
 
+unsigned long hda_get_max_frequency(struct snd_sof_dev *sdev)
```
I think you meant long long or u64, otherwise you have the ambiguity of unsigned long==u32 in some cases

ujfalusi replied:
No, I meant `unsigned long` to be aligned with the clock API in case  one wants to re-use this value in any way outside.
```c
unsigned long clk_get_rate(struct clk *clk);
int clk_set_rate(struct clk *clk, unsigned long rate);
```
and friends.

plbossart reviewed (changes requested):
still not getting the requirement to have ibs/obs always defined. Somehow I seem to be the only one lost with IPC4 concepts, but I clearly don't see how this will scale.

ranj063 commented:
@ujfalusi how do we plan to test this PR?

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1103,6 +1103,16 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+	if (!base_config->cpc) {
+		struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+
+		base_config->cpc = ipc4_data->max_core_frequency;
```
`base_config->cpc = ipc4_data->max_core_frequency / MSEC_PER_SEC;` or `/ 1000` as we want 1ms chunk number.

ujfalusi commented:
> @ujfalusi how do we plan to test this PR?

That's a good question, I think MTL have the manifest filled and @jsarha have the SOF PR for the topology override.

abonislawski commented:
@ujfalusi you can try mtl-004 branch (or upcoming mtl-005) with correct manifest data and clock selection based on this

ujfalusi committed:
ASoC: SOF: ipc4: Add SOF_IPC4_FW_PARAM_REGISTER_KCPS to fw params ID list

Add REGISTER_KCPS (22) to the firmware params id list, which is used via
LARGE_CONFIG_SET message for registering a new KCPS value which needs to
be taken into account by the firmware when selecting operating frequency.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4: Add sof_ipc4_send_kcps() function for sending kcps request

The sof_ipc4_send_kcps() can be used to register a new KCPS value
which needs to be taken into account by the firmware when selecting
operating frequency.
The kcps can be positive or negative number to force an increase in DSP
frequency or to allow taking lower speeds if allowed.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Add complete callback to sof_ipc_tplg_ops

Add optional ops to be called at the end of the topology loading.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: ipc4: Add custom kcontrol for 'DSP frequency boost'

Implement a simple switch which can be used to force the DSP to run in it's
highest frequency when enabled.

The underlying implementation uses the register kcps message to register or
unregister a kcps value equivalent to the DSP's highest operating
frequency.

The kcontrol is registered at the end of the topology loading in the
complete callback.

On the post firmware boot callback the boost is applied first if the boost
is enabled to make sure that it is taken into account right from the start.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: tokens: Move ACP I2S tokens to keep the list ordered

Move the ACP I2S tokens to keep the numbers in order.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v6:
- Move ACP tokens to keep the token numbers in order
- Add comments to clarify get_max_frequency (returns Hz) and the final fallback for CPC to use 1ms max frequency
- Add implementation of 'DSP frequency boost' via custom kcontrol switch to be used by user space to request full DSP speed.

ujfalusi commented:
Moving to draft for now.
New PRs might add subset of this PR depending on decisions on what and how differently we want to have the CPC/CPS support.

ujfalusi commented:
Changes since v7:
- No point of testing the boost flag at first boot
- Clean up the kcontrol put function

plbossart commented:
@ujfalusi this PR is replaced by #4338, yes?

ujfalusi committed:
Revert "ASoC: SOF: ipc4-topology: Do not use the CPC value from topology"

This reverts commit 68496f973adb5e783d14b468834f7fd785b2a057.

As part of the CPC value override, resurrect t he CPC value querry
from topology.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v8:
- rebased on the merged CPC code to avoid bitrotting this PR.

plbossart commented:
@ujfalusi can we close this?

ujfalusi commented:
> @ujfalusi can we close this?

If we are absolutely sure that a topology override is not going to be supported.

ujfalusi commented:
Changes since v9:
- rebased to resolve conflict (likely going to be closed)

plbossart commented:
> > @ujfalusi can we close this?
> 
> If we are absolutely sure that a topology override is not going to be supported.

For now the ask was to add support for CPC. It's upon the firmware team to clean-up the CPC values, we can only assume the information added to the manifest will be correct...

abonislawski commented:
Any chance to merge this?

plbossart commented:
> Any chance to merge this?

We already take the CPC from the manifest @abonislawski.

It's up to your team to provide 'correct' values for CPC, it's way too early to push code to the Linux kernel to add a workaround. Let's fix the manifests first.

ujfalusi commented:
I guess, this can be closed...
